### PR TITLE
Adding support for filtering whatif output

### PR DESCRIPTION
**Pull request by jasonbrisbin**
## Overview/Summary
Adding support for filtering whatif results using the ExcludeChangeType parameter for the Get-Az****DeploymentWhatIfResult cmdlets.  This is in response to issue 475 that I opened.  I propose using a PSF config value for controlling this i.e. AzOps.Core.WhatifExcludedChangeTypes which would be an array containing the list of change types to exclude.  https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deploy-what-if?tabs=azure-powershell#change-types

Also I am suggesting a slight change to the formatting of output.json such that the output will be a valid JSON document when there is more than 1 template deployed with changes.  Currently the file has a simple text append function which when processing multiple templates in a single run will produce a ill formatted document. 

Before
```json
{
  "results":"first template"
}
{
  "results":"second template"
}
```

After
```json
[
  {
    "results":"first template"
  },
  {
    "results":"second template"
  }
]
```
Replace this with a brief description of what this Pull Request fixes, changes, etc.

## This PR fixes/adds/changes/removes

1. https://github.com/Azure/AzOps/issues/475

### Breaking Changes
1. None

## Testing Evidence

Please provide any testing evidence to show that your Pull Request works/fixes as described and planned (include screenshots, if appropriate).

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/azops/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/azops/issues), for tracking and closure.
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/azops/tree/main)
- [x] Performed testing and provided evidence.
- [x] Updated relevant and associated documentation.
- [ ] Updated the ["What's New?"](https://github.com/Azure/Enterprise-Scale/wiki/Whats-new) wiki page (located in the [Enterprise-Scale repo](https://github.com/Azure/Enterprise-Scale) in the directory: `/docs/wiki/whats-new.md`)

Settings.json with whatif exclusions for NoChange and Ignore
image

Post change sample output with obfuscated names/guids.  This is the output.json, note the Change Type field and the top level array format of the document.
[sample.txt](https://github.com/Azure/AzOps/files/7565849/sample.txt)

### CODE REVIEW

**Review by daltondhcp — CHANGES_REQUESTED** [review]
Thank you for your contribution @jasonbrisbin. 

Based on what is currently in the PR, it doesn't look like you have touched the MD file at all (which is what is added to the PR comment). https://github.com/Azure/AzOps-Accelerator/blob/main/.github/workflows/validate.yml#L115 
It also looks like we potentially have bit of a clash with #470 that was merged yesterday

Anyway, to move forward with this, can you please do the following:
1. Share examples (screenshots) of what the output would look like so we understand how it helps to highlight what has been changed and reduce noise. 
2. Validate the end-to-end flow in a pipeline/GitHub actions as per https://github.com/azure/azops/wiki/pre-release. 
3. Update https://github.com/Azure/AzOps/blob/main/src/internal/configurations/Core.ps1 and https://github.com/Azure/AzOps-Accelerator/blob/main/settings.json with the new setting/suggested default values. 
4. Update https://github.com/azure/azops/wiki/settings so the we have the new setting documented

**Comment by jasonbrisbin** [comment]
@Jefajers I reran today with the published changes from 470 and I really like the output it does a good job cleaning up the output and making it more human readable. 

image


@daltondhcp  That said, it would be ideal if the json file produced output was valid for multiple deploys.  Here is a sanitized output.json from my most recent run.  Notice that there are multiple independent JSON objects at the top level and no commas between them (lines 61/62,122/123,183/184).  My update adds a top level array and each json object is a member of that array.
[OUTPUT.txt](https://github.com/Azure/AzOps/files/7571779/OUTPUT.txt)

**Comment by daltondhcp** [comment]
@jasonbrisbin , so you'd want to keep the new MD output, but change/enable the new feature for the output.json so you can use that for other purposes in the pipeline? Can you share your use case in that case?

2 is essentially the ability to install the module from your fork/branch. See example how I use it in my repo @ https://github.com/daltondhcp/azops-test2/blob/main/.github/workflows/pull.yml#L168 .

**Comment by jasonbrisbin** [comment]
Absolutely, we want to keep the changes from #470.  My update should compliment that change.  I specifically was trying to accomplish the following. 
1.  Ensure output.json is a well formatted document.  As it stands today, I am not sure the current ouput.json is useful when dealing with more than one template being deployed.
2.  Adds support for excluding specific change types from the whatif output.  The format of the output is great but it still includes all possible results.  Below is an excerpt from this article

https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deploy-what-if?tabs=azure-powershell#change-types
## Change types
The what-if operation lists six different types of changes:

**Create**: The resource doesn't currently exist but is defined in the template. The resource will be created.

**Delete**: This change type only applies when using complete mode for deployment. The resource exists, but isn't defined in the template. With complete mode, the resource will be deleted. Only resources that support complete mode deletion are included in this change type.

**Ignore**: The resource exists, but isn't defined in the template. The resource won't be deployed or modified.

**NoChange**: The resource exists, and is defined in the template. The resource will be redeployed, but the properties of the resource won't change. This change type is returned when ResultFormat is set to FullResourcePayloads, which is the default value.

**Modify**: The resource exists, and is defined in the template. The resource will be redeployed, and the properties of the resource will change. This change type is returned when ResultFormat is set to FullResourcePayloads, which is the default value.

**Deploy**: The resource exists, and is defined in the template. The resource will be redeployed. The properties of the resource may or may not change. The operation returns this change type when it doesn't have enough information to determine if any properties will change. You only see this condition when ResultFormat is set to ResourceIdOnly.

I will setup my pipeline like yours to complete #2.  

Thanks @daltondhcp

**Comment by daltondhcp** [comment]
@jasonbrisbin, thank you for the detailed explanation! Looking forward to your updated PR 👍

**Comment by daltondhcp** [comment]
Hey @jasonbrisbin - have you had the time to take this any further?

Thanks in advance!

**Comment by jasonbrisbin** [comment]
I was out last week, but will be working on it this week.

**Comment by jasonbrisbin** [comment]
@daltondhcp, I copied your pull pipeline and it runs successfully as shown here

https://github.com/vancopayments/AzOps/actions/runs/1527490982

This should address item 2.

**Comment by daltondhcp** [comment]
Sorry for the delay @jasonbrisbin - I have been on leave for almost two weeks - will take a look and get back to you around this during the week.

File: src/internal/functions/New-AzOpsDeployment.ps1 [hunk 1]
```diff
@@ -146,8 +150,12 @@
                 if ($TemplateParameterFilePath) {
                     $parameters.TemplateParameterFile = $TemplateParameterFilePath
                 }
-
+                $WhatifExcludedChangeTypes = Get-PSFConfigValue -FullName 'AzOps.Core.WhatifExcludedChangeTypes'
+                if($WhatifExcludedChangeTypes){
+                    $parameters.ExcludeChangeType = $WhatifExcludedChangeTypes
+                }
                 $results = Get-AzResourceGroupDeploymentWhatIfResult @parameters -ErrorAction Continue -ErrorVariable resultsError
+                if($parameters.ExcludeChangeType){$parameters.Remove('ExcludeChangeType')}
```
Comment (daltondhcp) [inline src/internal/functions/New-AzOpsDeployment.ps1:158]:
```suggestion
                if ($parameters.ExcludeChangeType) {
                    $parameters.Remove('ExcludeChangeType')
                }
```

**Review by daltondhcp — CHANGES_REQUESTED** [review]
Looks good, just some cosmetic changes needed to ensure we are aligned with rest of the code base.

**Comment by jasonbrisbin** [comment]
> Looks good, just some cosmetic changes needed to ensure we are aligned with rest of the code base.

@daltondhcp  I have made the changes you suggested.

File: src/internal/functions/New-AzOpsDeployment.ps1 [hunk 2]
```diff
@@ -45,7 +45,11 @@
         $Mode = "Incremental",
 
         [string]
-        $StatePath = (Get-PSFConfigValue -FullName 'AzOps.Core.State')
+        $StatePath = (Get-PSFConfigValue -FullName 'AzOps.Core.State'),
+
+        [string]
```
Comment (daltondhcp) [inline src/internal/functions/New-AzOpsDeployment.ps1:50]: Please change to array string
```suggestion
        [string[]]        
```

File: src/internal/functions/New-AzOpsDeployment.ps1 [hunk 3]
```diff
@@ -100,9 +104,12 @@
                 if ((Get-AzContext).Subscription.Id -ne $scopeObject.subscription) {
                     Set-AzOpsContext -ScopeObject $scopeObject
                 }
-
+                if ($WhatifExcludedChangeTypes) {
+                    $parameters.ExcludeChangeType = $WhatifExcludedChangeTypes
+                }
                 # Validate Template
                 $results = Get-AzSubscriptionDeploymentWhatIfResult @parameters -ErrorAction Continue -ErrorVariable resultsError
+                if($parameters.ExcludeChangeType){$parameters.Remove('ExcludeChangeType')}
```
Comment (daltondhcp) [inline src/internal/functions/New-AzOpsDeployment.ps1:112]: Cosmetic change so it is consistent :) 
```suggestion
                if ($parameters.ExcludeChangeType) {
                    $parameters.Remove('ExcludeChangeType')
                }
```

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -1,5 +0,0 @@
-# See for instructions on this file https://help.github.com/articles/about-codeowners/
```
Comment (daltondhcp) [inline .github/CODEOWNERS:1]: Please remove codeowners from the PR, we still need it :)

File: .gitignore [hunk 5]
```diff
@@ -33,4 +33,4 @@ root/*
 
 # Debug
 **.private**
-debug/*
```
Comment (daltondhcp) [inline .gitignore:36]: Please remove .gitignore from the PR

File: src/internal/functions/Set-AzOpsWhatIfOutput.ps1 [hunk 6]
```diff
@@ -45,7 +45,15 @@
             else {
                 $mdOutput = 'WhatIf Results: Resource Creation:{0}```{0}{1}{0}```{0}' -f [environment]::NewLine, $resultString
             }
-            Add-Content -Path '/tmp/OUTPUT.json' -Value $resultJson -WhatIf:$false
+            $existingContent = @(Get-Content -Path '/tmp/OUTPUT.json' -Raw | ConvertFrom-Json)
+            if ($existingContent.count -gt 0) {
+                $existingContent += $results.Changes
+                $existingContent = $existingContent|ConvertTo-Json -Depth 100
```
Comment (daltondhcp) [inline src/internal/functions/Set-AzOpsWhatIfOutput.ps1:51]:
```suggestion
                $existingContent = $existingContent | ConvertTo-Json -Depth 100
```

**Review by daltondhcp — CHANGES_REQUESTED** [review]
A few more things, thank you for your efforts and patience! :)

File: .github/CODEOWNERS (hunk 4 above)
Comment (jasonbrisbin) [reply .github/CODEOWNERS:1]: I believe this is done, can you rereview @daltondhcp

Comment (daltondhcp) [reply .github/CODEOWNERS:1]: Both of the files are removed entirely in your branch, meaning that they would be removed in upstream at merge, which we obviously don't want. Please restore CODEOWNERS and .gitignore in your branch from Azure/AzOps:main and push. 
image

**Review by daltondhcp — APPROVED** [review]

**Pull request merged**